File: Recursion-Backtracking/08-Permutations/permutations.cpp
class Solution {
public:

    void findPermutation(int index, vector<int>& nums, vector<vector<int>>& result){
        if(index == nums.size()){
            result.push_back(nums);
            return;
        }

        for(int i = index ; i < nums.size(); i++){
            swap(nums[index], nums[i]);
            findPermutation(index + 1, nums, result);
            swap(nums[index], nums[i]);
        }
    }

    vector<vector<int>> permute(vector<int>& nums) {
         vector<vector<int>> result;
         findPermutation(0, nums, result);
         return result;
    }
};

File: Recursion-Backtracking/08-Permutations/permutations.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */

function findPermutations(index, nums, result){
    if(index === nums.length){
        result.push([...nums]);
        return;
    }

    for(let i = index ; i < nums.length; i++){
        [nums[i], nums[index]] = [nums[index], nums[i]];
        findPermutations(index + 1, nums, result);
        [nums[i], nums[index]] = [nums[index], nums[i]];
    }
}

var permute = function(nums) {
    let result = [];
    findPermutations(0, nums, result);
    return result;
};